include/seahorn/CexExeGenerator.hh

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Bitcode/BitcodeWriter.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/ValueMap.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/ToolOutputFile.h"
+
+#include "boost/algorithm/string/replace.hpp"
+
+#include "seahorn/Bmc.hh"
+#include "seahorn/Expr/ExprLlvm.hh"
+#include "seahorn/Expr/ExprMemMap.hh"
+#include "seahorn/Expr/ExprOpBinder.hh"
+#include "seahorn/MemSimulator.hh"
+#include "seahorn/SolverBmc.hh"
+#include "seahorn/Support/SeaDebug.h"
+#include "seahorn/Transforms/Instrumentation/ShadowMemDsa.hh"
+#include <memory>
+
+namespace llvm {
+class TargetLibraryInfo;
+class DataLayout;
+class LLVMContext;
+class Module;
+} // namespace llvm
+
+/*
+    Converts bmc trace into a linkable LLVM IR
+    with all non-det functions implemented
+ */
+
+namespace seahorn {
+namespace cexGen {
+
+using MemMap = expr::exprMemMap::ExprMemMap;
+using DsaKeyVal = std::map<Expr, Expr>;
+
+// helper functions
+namespace utils {
+
+/// \brief converts Expr \p e into LLVM Constant
+Constant *exprToLlvm(Type *ty, Expr e, LLVMContext &ctx, const DataLayout &dl);
+
+/// \brief extract default value and special id-val pairs from Expr encoding of
+/// a shadow mem segment \p segment using ExprMemMap; returns a ConstantArray
+/// of length \p size divided by content width, filled with default and special
+/// values
+Constant *exprToMemSegment(Expr segment, Expr startAddr, Expr size,
+                           llvm::LLVMContext &ctx, const llvm::DataLayout &dl);
+
+/// \brief extract content from array expression \p e using ExprMemMap;
+/// if sucessful, store default to \p e defaultValue, special id/value pairs
+/// to \p out
+bool extractArrayContents(Expr e, DsaKeyVal &out, Expr &defaultValue);
+
+} // namespace utils
+
+template <class Trace> class CexExeGenerator {
+  Trace &m_trace;
+  const DataLayout &m_dl;
+  const TargetLibraryInfo &m_tli;
+  LLVMContext &m_context;
+  std::unique_ptr<Module> m_harness;
+
+  // map function calls to return value(s)
+  ValueMap<const Function *, ExprVector> m_func_val_map;
+  // map a dsa node to start and end addresses
+  std::map<unsigned, std::pair<Expr, Expr>> m_dsa_alloc_map;
+  // map a dsa node to its contents (as a pair of a default value
+  // and a map from index to value)
+  std::map<unsigned, std::pair<Expr, DsaKeyVal>> m_dsa_content_map;
+  // list of <start addr, size> of havoc pointers
+  std::vector<std::pair<Expr, Expr>> m_havoc_ptrs;
```

Comment:
what are havoc pointers?

Reply:
start address and size of each `memhavoc` invocation, maybe change to something like `m_memhavoc_args`?